### PR TITLE
refactor(core): Move active workflows endpoints to a decorated controller class (no-changelog)

### PR DESCRIPTION
**Pull request by netroy**
## Summary
This is a continuation of migrating all rest endpoints to decorated controller classes

## Review / Merge checklist
- [x] PR title and summary are descriptive
- [x] Tests included

### CODE REVIEW

File: packages/cli/src/controllers/activeWorkflows.controller.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	}
+
+	@Get('/error/:id')
+	async getActiveError(req: WorkflowRequest.GetActivationError) {
+		const {
+			user,
+			params: { id: workflowId },
+		} = req;
+		return this.activeWorkflowsService.getActivationError(workflowId, user);
+	}
```
Comment (netroy) [inline packages/cli/src/controllers/activeWorkflows.controller.ts:24]: TODO: add integration tests

File: packages/cli/src/services/activeWorkflows.service.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+import type { User } from '@db/entities/User';
+import { SharedWorkflowRepository } from '@db/repositories/sharedWorkflow.repository';
+import { WorkflowRepository } from '@db/repositories/workflow.repository';
+import { ActivationErrorsService } from '@/ActivationErrors.service';
+import { BadRequestError } from '@/errors/response-errors/bad-request.error';
+import { Logger } from '@/Logger';
+
+@Service()
+export class ActiveWorkflowsService {
```
Comment (netroy) [inline packages/cli/src/services/activeWorkflows.service.ts:11]: TODO: Add unit tests with 100% coverage

File: packages/cli/src/controllers/activeWorkflows.controller.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { Service } from 'typedi';
+import { Authorized, Get, RestController } from '@/decorators';
+import { WorkflowRequest } from '@/requests';
+import { ActiveWorkflowsService } from '@/services/activeWorkflows.service';
+
+@Service()
+@Authorized()
+@RestController('/active_workflows')
```
Comment (ivov) [inline packages/cli/src/controllers/activeWorkflows.controller.ts:8]:
```suggestion
@RestController('/active-workflows')
```

File: cypress/e2e/30-editor-after-route-changes.cy.ts [hunk 4]
```diff
@@ -199,7 +199,7 @@ describe('Editor zoom should work after route changes', () => {
 		cy.intercept('GET', '/rest/workflow-history/workflow/*').as('getHistory');
 		cy.intercept('GET', '/rest/users').as('getUsers');
 		cy.intercept('GET', '/rest/workflows').as('getWorkflows');
-		cy.intercept('GET', '/rest/active').as('getActive');
+		cy.intercept('GET', '/rest/active_workflows').as('getActiveWorkflows');
```
Comment (ivov) [inline cypress/e2e/30-editor-after-route-changes.cy.ts:202]:
```suggestion
		cy.intercept('GET', '/rest/active-workflows').as('getActiveWorkflows');
```

File: packages/cli/src/controllers/activeWorkflows.controller.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+export class ActiveWorkflowsController {
+	constructor(private readonly activeWorkflowsService: ActiveWorkflowsService) {}
+
+	@Get('/')
+	async getActiveWorkflows(req: WorkflowRequest.GetAllActive) {
+		return this.activeWorkflowsService.getAllActiveIdsForUser(req.user);
+	}
+
+	@Get('/error/:id')
+	async getActiveError(req: WorkflowRequest.GetActivationError) {
```
Comment (ivov) [inline packages/cli/src/controllers/activeWorkflows.controller.ts:18]:
```suggestion
	async getActivationError(req: WorkflowRequest.GetActivationError) {
```

File: packages/cli/src/databases/repositories/sharedWorkflow.repository.ts [hunk 6]
```diff
@@ -1,10 +1,31 @@
...
+
+	async hasAccess(workflowId: string, user: User) {
+		const where: FindOptionsWhere<SharedWorkflow> = {
+			workflowId,
+		};
+		if (!user.hasGlobalScope('workflow:read')) {
+			where.userId = user.id;
+		}
+		return this.exist({ where });
+	}
```
Comment (ivov) [inline packages/cli/src/databases/repositories/sharedWorkflow.repository.ts:20]: I'd find it more intuitive if `hasAccess` were defined on the `User` entity, just like `user.hasGlobalScope()`.

File: packages/cli/src/services/activeWorkflows.service.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+		private readonly activationErrorsService: ActivationErrorsService,
+	) {}
+
+	async getAllActiveIds() {
+		const activationErrors = await this.activationErrorsService.getAll();
+		const activeWorkflowIds = await this.workflowRepository.getActiveIds();
+		return activeWorkflowIds.filter((workflowId) => !activationErrors[workflowId]);
+	}
+
+	async getAllActiveIdsForUser(user: User) {
```
Comment (ivov) [inline packages/cli/src/services/activeWorkflows.service.ts:25]: Nit - Already clear from the arg name and type.

```suggestion
	async getAllActiveIdsFor(user: User) {
```

File: packages/cli/src/services/activeWorkflows.service.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+	async getActivationError(workflowId: string, user: User) {
+		const hasAccess = await this.sharedWorkflowRepository.hasAccess(workflowId, user);
+		if (!hasAccess) {
+			this.logger.verbose('User attempted to access workflow errors without permissions', {
+				workflowId,
+				userId: user.id,
+			});
+
+			throw new BadRequestError(`Workflow with ID "${workflowId}" could not be found.`);
```
Comment (ivov) [inline packages/cli/src/services/activeWorkflows.service.ts:47]: For learning - I'd've used at `BadRequestError` and other response errors only at controller level. It's okay to use them at service level?

File: packages/core/src/ActiveWorkflows.ts [hunk 9]
```diff
@@ -22,10 +23,9 @@ import {
 
 import type { IWorkflowData } from './Interfaces';
 
+@Service()
 export class ActiveWorkflows {
-	private activeWorkflows: {
-		[workflowId: string]: IWorkflowData;
-	} = {};
+	private activeWorkflows: Record<string, IWorkflowData> = {};
```
Comment (ivov) [inline packages/core/src/ActiveWorkflows.ts:28]: Same comment as above about preserving key name for readability.

File: packages/editor-ui/src/api/workflows.ts [hunk 10]
```diff
@@ -24,7 +24,7 @@ export async function getWorkflows(context: IRestApiContext, filter?: object) {
 }
 
 export async function getActiveWorkflows(context: IRestApiContext) {
-	return makeRestApiRequest(context, 'GET', '/active');
+	return makeRestApiRequest(context, 'GET', '/active_workflows');
```
Comment (ivov) [inline packages/editor-ui/src/api/workflows.ts:27]:
```suggestion
	return makeRestApiRequest(context, 'GET', '/active-workflows');
```

File: packages/editor-ui/src/stores/workflows.store.ts [hunk 11]
```diff
@@ -373,7 +373,11 @@ export const useWorkflowsStore = defineStore(STORES.WORKFLOWS, {
 
 		async getActivationError(id: string): Promise<string | undefined> {
 			const rootStore = useRootStore();
-			return makeRestApiRequest(rootStore.getRestApiContext, 'GET', `/active/error/${id}`);
+			return makeRestApiRequest(
+				rootStore.getRestApiContext,
+				'GET',
+				`/active_workflows/error/${id}`,
```
Comment (ivov) [inline packages/editor-ui/src/stores/workflows.store.ts:379]:
```suggestion
				`/active-workflows/error/${id}`,
```

File: packages/cli/src/controllers/activeWorkflows.controller.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import { Service } from 'typedi';
+import { Authorized, Get, RestController } from '@/decorators';
+import { WorkflowRequest } from '@/requests';
```
Comment (ivov) [inline packages/cli/src/controllers/activeWorkflows.controller.ts:3]: I'd've expected this to be linted:

```suggestion
import type { WorkflowRequest } from '@/requests';
```

**Review by ivov — COMMENTED** [review]
Nice work! Only minor comments.

File: packages/cli/src/ActiveWorkflowRunner.ts [hunk 13]
```diff
@@ -57,44 +56,40 @@ import {
...
-
-	private queuedActivations: {
-		[workflowId: string]: {
-			activationMode: WorkflowActivateMode;
-			lastTimeout: number;
-			timeout: NodeJS.Timeout;
-			workflowData: IWorkflowDb;
-		};
-	} = {};
+	private queuedActivations: Record<string, QueuedActivation> = {};
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/cli/src/ActiveWorkflowRunner.ts:78]: How about `Record<WorkflowEntity['id'], QueuedActivation>` instead ?

File: packages/cli/src/controllers/activeWorkflows.controller.ts (hunk 12 above)
Comment (netroy) [reply packages/cli/src/controllers/activeWorkflows.controller.ts:3]: this is "required" for decorator metadata, even though this is just a type, and doesn't actually add anything in the compiled code.
in general, we should worry about this stuff in reviews anymore, and let the `consistent-type-imports` rule take care of it.

File: packages/cli/src/services/activeWorkflows.service.ts (hunk 8 above)
Comment (netroy) [reply packages/cli/src/services/activeWorkflows.service.ts:47]: I have conflicting opinions about this. one one hand all communication layer details should be abstracted away in the controllers, on the other hand letting services throw these error makes the code a lot simpler and easier to maintain 🤷🏽.

File: packages/cli/src/databases/repositories/sharedWorkflow.repository.ts (hunk 6 above)
Comment (netroy) [reply packages/cli/src/databases/repositories/sharedWorkflow.repository.ts:20]: that would be nice, but doing so in the current code could create a circular dependency between entities and repositories. I'll think about it, and try to find alternative ways of doing this that feel more intuitive.

File: packages/cli/src/controllers/activeWorkflows.controller.ts (hunk 12 above)
Comment (ivov) [reply packages/cli/src/controllers/activeWorkflows.controller.ts:3]: > we should worry about this stuff in reviews anymore

Definitely, it just caught my eye and I got curious to know why this type was exempted :)

**Review by ivov — APPROVED** [review]
🏅

**Comment by Joffcom** [comment]
Got released with [`n8n@1.23.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.23.0)

**Pull request merged**